apisix/secret/aws.lua

```diff
@@ -0,0 +1,193 @@
+--
+
+--- AWS Tools.
+local core = require("apisix.core")
+local http = require("resty.http")
+local aws = require("resty.aws")
+
+local sub = core.string.sub
+local rfind_char = core.string.rfind_char
+local env = core.env
+local type = type
+local unpack = unpack
+
+--- AWS Environment Configuration
+local AWS
+local AWS_ACCESS_KEY_ID
+local AWS_SECRET_ACCESS_KEY
+local AWS_SESSION_TOKEN
+local AWS_REGION
+
+local schema = {
+    type = "object",
+    properties = {
+        access_key_id = {
+            type = "string",
+        },
+        secret_access_key = {
+            type = "string",
+        },
+        session_token = {
+            type = "string",
+        },
+        region = {
+            type = "string",
+        },
+        endpoint_url = core.schema.uri_def,
+    },
+}
+
+local _M = {
+    schema = schema
+}
+
+local function initialize_aws()
+    AWS_ACCESS_KEY_ID = env.fetch_by_uri("$ENV://AWS_ACCESS_KEY_ID")
+    AWS_SECRET_ACCESS_KEY = env.fetch_by_uri("$ENV://AWS_SECRET_ACCESS_KEY")
+    AWS_SESSION_TOKEN = env.fetch_by_uri("$ENV://AWS_SESSION_TOKEN")
+    AWS_REGION = env.fetch_by_uri("$ENV://AWS_REGION")
+    AWS = aws()
+    initialize_aws= nil
+end
+
+local function make_request_to_aws(conf,key)
+    if initialize_aws then
+        initialize_aws()
+    end
+
+    local region = conf.region or AWS_REGION
+    if not region then
+        return nil, "aws secret manager requires region"
+    end
+
+    local access_key_id = env.fetch_by_uri(conf.access_key_id)
+    if not access_key_id then
+        access_key_id = conf.access_key_id or AWS_ACCESS_KEY_ID
+    end
+
+    local secret_access_key = env.fetch_by_uri(conf.secret_access_key)
+    if not secret_access_key then
+        secret_access_key = conf.secret_access_key or AWS_SECRET_ACCESS_KEY
+    end
+
+    local session_token = env.fetch_by_uri(conf.session_token)
+    if not session_token then
+        session_token = conf.session_token or AWS_SESSION_TOKEN
+    end
+
+    local my_creds = nil
+    if access_key_id and secret_access_key then
+        my_creds = AWS:Credentials {
+            accessKeyId = access_key_id,
+            secretAccessKey = secret_access_key,
+            sessionToken = session_token,
+        }
+    end
+
+    if not my_creds then
+        return nil, "(invalid credentials)"
+    end
+
+    AWS.config.credentials = my_creds
+
+    local default_endpoint = "https://secretsmanager." .. region .. ".amazonaws.com"
+    local pre, host, port, _, _ = unpack(http:parse_uri(conf.endpoint_url or default_endpoint))
+    local endpoint = pre .. "://" .. host
+
+    local sm = AWS:SecretsManager {
+        endpoint = endpoint,
+        region = region,
+        port = port,
+    }
+
+    local res, err = sm:getSecretValue {
+        SecretId = key,
+        VersionStage = "AWSCURRENT",
+    }
+
+    if type(res) ~= "table" then
+        if err then
+            return nil, err
+        end
+        return nil, "(invalid response)"
```

Comment:
we need to log error info

Reply:
replace by https://github.com/apache/apisix/pull/11417#discussion_r1714764673